libexec/mass-testing/test-harness

```diff
@@ -0,0 +1,235 @@
+    root_process = psutil.Process(pid)
+    if include_root:
+        pool = [root_process]
+    else:
+        pool = [*root_process.children()]
+
+    while len(pool) > 0:
+        new_pool = []
+        for element in pool:
+            yield element
+            new_pool.extend(element.children())
+        pool = new_pool
+
+
+def kill_and_reap_children(exclude_pids: Collection[int]):
+    # Kill all child processes and reap them. Returns True if any process was
+    # forcefully killed.
+    #
+    # NOTE: this creates a race, as new processes might be spawned in the
+    # meantime. An alternative might be to stuff the processes in a cgroupv2
+    # and use the 'cgroup.kill' special file
+    process_killed = False
+    while True:
+        children = [
+            p
+            for p in reversed(list(walk_process_tree(os.getpid(), False)))
+            if p.pid not in exclude_pids
+        ]
+        if len(children) == 0:
+            break
+
+        for process in children:
+            status = None
+            ppid = None
+            with suppress(psutil.NoSuchProcess):
+                status = process.status()
+            with suppress(psutil.NoSuchProcess):
+                ppid = process.ppid()
+
+            if status is None or ppid is None:
+                continue
+
+            if status not in (psutil.STATUS_ZOMBIE, psutil.STATUS_DEAD):
+                kill(process.pid, signal.SIGABRT)
+                process_killed = True
+                # Wait a bit between kills to allow for things to settle
+                sleep(2)
+            elif ppid == os.getpid():
+                os.waitpid(process.pid, 0)
+    return process_killed
+
+
+def pid_has_signal(pid: int, signalno: int) -> bool:
+    signal_bits = 0
+    with open(f"/proc/{pid}/status") as file:
+        for line in file:
+            if line.startswith("SigCgt:"):
+                signal_bits = int(line.removeprefix("SigCgt:"), 16)
+    return signal_bits & (2 ** (signalno - 1)) > 0
+
+
+def handle_timeout(root_process):
+    main_pid = None
+    for process in walk_process_tree(root_process.pid):
+        if pid_has_signal(process.pid, signal.SIGABRT):
+            main_pid = process.pid
+            break
+
+    if main_pid is None:
+        # Bail
+        root_process.send_signal(signal.SIGINT)
+        return
+
+    kill(main_pid, signal.SIGABRT)
+    try:
+        process.wait(30)
+    except psutil.TimeoutExpired:
+        kill(main_pid, signal.SIGKILL)
+        process.wait()
+    except psutil.ChildProcessError:
+        # Process died before we could `wait`
+        pass
+
+
+def patch_cmd(original_cmd: List[str], input_: str | None) -> List[str]:
+    result = []
+    for part in original_cmd:
+        if part == input_:
+            result.append("$INPUT")
+        elif Path(part).resolve() == (SCRIPT_DIR / "revng-test"):
+            result.append("revng")
```

Comment:
This script must not assume anything `revng` specific. Please double check.

Reply:
Ok, but then this makes creating a reproducible `commands.json` impossible. Copying `argv` 1:1 is not pratical.

Reply:
Elaborate, like "if we don't do this, this specifically will happen".

Reply:
Basically, in order to have a proper `reproducer.tar`, it's needed to store somewhere something like `argv`. But if the argv was copied 1:1 you would have:
```
test-harness --output <some directory on the build machine> --input <some path on the build machine> ... -- revng-test <options> %INPUT%
```

In an ideal world, the reproducer script would call:
```
revng <options> %INPUT%
```

So that none of the `revng-test` logic get in the way of the reproducer.

Reply:
Fixed by:
* Using `%INPUT%` so that it can be easily replaced
* Having `hooks.pre` and `hooks.post` in `meta.yml` to mangle `$PATH` so that `revng` is a symlink to `revng-test`